### ISSUE TITLE
Learning New IP Addresses

### ISSUE THREAD

**Issue by nibanks** [nibanks/quic-new-path#5]
Related to #2, the mechanisms there would allow you to discover the "public" IP address of the currently used addresses, but if the goal is to switch to a different local address, you don't know what the peer would see for that address. And, if the peer's NAT/firewall won't let your packet through, until they've already sent a packet back to you, you've got a "chicken and the egg" problem.

Take the following (worst case?) example:

Server `S`:
 - Has a single IP address, `S1` on a private (corporate) network (`N`) that is exposed via a VPN.
 - Has access to the public internet via outgoing traffic.
 - Access to the internet goes through a (symmetric) NAT/Firewall that dynamically uses different ports **and** IP addresses externally.

Client `C`:
 - Has a single IP address, `C1` on a private (home) network.
 - Has a VPN connection to `N`, exposed locally as address `C2`.
 - Has access to the public internet via outgoing traffic.
 - Access to the internet goes through a NAT.

`C` initially connects to `S` from `C2` to `S1`. Neither address is NAT'ed and the connection succeeds. So both sides natively see the "public" address being used. The great complexity then comes when the client wants to natively use `C1` to `S1` though `S`'s firewall (not VPN).

Step 1, the client has to be to discover its public address. Assuming the extension adds the ability to get this from any path challenge/response, then this could be discovered by connecting to any publicly accessible QUIC server that supports the extension and getting it from there. If the extension is widely adopted then picking a server shouldn't be a problem. If not, we'd have to set up some dedicated/well-known server that supports it. But either way, it would require a separate QUIC connection to a new server to discover. One question here would be if this should all happen automatically under the covers from the app, or if the app should be required to make that new connection. I'd rather hide the complexity from the app. I'd probably implement it as a global (to the library) cache, and expose it as an async request cache look up that an existing connection can make.

Step 2, assuming the client now knows its public address (`C1_P`) for `C1` it can send the new path request to the server with that  address as the origin. Additionally, it starts listening for incoming packets for the connection on `C1` now, in addition to `C2`.

Step 3, the server gets the request and accepts. It should respond with the address of its own; going through a similar process to that of the client. **But** the problem here is that if/when the server probes for its public address, it will be different (definitely port, and most like IP address) than the one that ends up being used for contacting the client. I know of no existing and widely adopted protocol that allows an endpoint to get from the network (or NAT) "What's my IP address when I try to contact X:Y?". So, the server ends up getting **an** address (`S1_P0`), which won't end up being useful here.

Step 4, the server sends the new path response to `C2` with `S1_P0` as its origin address. Then, (after a short delay?) it begins probing/punching through to `C1_P`.

Step 5a, the client receives the new path response on `C2`. It then immediately begins probing/punching through to `S1_P0`.

Step 5b, the client ideally receives the probes from `S` (possibly even before 5a), but on address `S1_P1`, finally discovering the true address of the server. Assuming the packet can be decrypted and processed successfully, the client then adds/uses this path just as defined in the base QUIC spec. It responds on it with a path challenge. The client should then abandon the path to `S1_P0`.

Step 6, the server receives the clients response, and everything progresses as defined in the base QUIC spec.

**Now**, for the problem. Practically, if the client has a symmetric NAT, things just aren't going to work without additional logic. My home network has just this setup. For me to get things working, I had to throw UPnP in the mix. In addition, on Step 2, the client would have to instruct the NAT to allow all traffic (thus completely opening the public port, temporarily) on the public IP:port combination. If the client was able to discover the server's real IP then that wouldn't be necessary.

**Comment by huitema** [comment]
In the general case, you are saying that this should work if one of the peers knows its real address, even if the other is behind a symmetric NAT and does not. Assume a simple case, in which the peer that "knows" its address is behind a stateful firewall. It knows its own address, and may communicate that to its peer through the VPN. But it needs to "guess" the other peer's address in order to open the stateful firewall, and that leads to a lot of complexity.

**Comment by nibanks** [comment]
Makes sense. I think PCP (https://datatracker.ietf.org/doc/rfc6887/) would be a good thing to use for this scenario. I just validated and at least on my home network it works. I could replace UPnP static mappings with a PCP MAP. It also looks pretty easy to implement in a cross platform library. Another nice thing about PCP, it seems to indicate the public IP address if it works; so there's no need to "discover" that address manually.

Also worth noting, I tried it on my corporate network and it did not work.

**Comment by huitema** [comment]
Yes, PCP makes sense. Just need a bit of smarts to demux PCP and QUIC. I guess a little bit of experimenting will help.

By the way, if I were you, I would not start hacking a corporate firewall without giving a heads up to my boss...
